adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -289,12 +292,26 @@ private void doUploadMessage(final AmqpContext context, final Future<MessageSend
             }
             return Future.<Void> succeededFuture();
         }).recover(t -> {
-            LOG.info("Cannot process message for Device [deviceId: {}] with Tenant [tenantId: {}]",
+            LOG.debug("Cannot process message for Device [tenantId: {}, deviceId: {}, endpoint: {}]",
+                    context.getTenantId(),
                     context.getDeviceId(),
-                    context.getTenantId());
-            if (!context.remotelySettled()) {
+                    context.getEndpoint());
+            if (!context.isRemotelySettled()) {
                 // client wants to be informed that the message cannot be processed.
-                context.released();
+
+                // t can only be a ServiceInvocationException
+                final ServiceInvocationException error = (ServiceInvocationException) t;
+                final StringBuilder errorMsg = new StringBuilder();
+
+                if (ServerErrorException.class.isInstance(t)) {
+                    errorMsg.append("Releasing message: {} - {}");
+                    context.released();
+                } else {
+                    errorMsg.append("Rejecting message : {} - {}");
+                    context.reject(AmqpError.NOT_FOUND, (ClientErrorException) error);
+                }
+
+                LOG.debug(errorMsg.toString(), error.getErrorCode(), error.getMessage());
```

Comment:
please create only one log statement for the situation. Using multiple statements will make it close to impossible to correlate them if we are receiving more than one message per second ...